src/viewport/index.tsx

```diff
@@ -596,12 +595,15 @@ class Viewport extends React.Component<
         }
     }
 
-    public changeAgentsColor(colorChange: ColorChange): void {
-        const { agent, color } = colorChange;
-        const agentIds = this.selectionInterface.getAgentIdsByNamesAndTags([
-            agent,
-        ]);
-        this.visGeometry.applyColorToAgents(agentIds, color);
+    public changeAgentsColor(appliedColors: UIDisplayData): void {
+        const colorSettings =
+            this.selectionInterface.getIdsAndColorsFromUIData(appliedColors);
+        colorSettings.forEach((setting) => {
+            this.visGeometry.applyColorToAgents(
+                setting.agentIds,
+                setting.color
+            );
+        });
```

Comment:
This function now takes in a snapshot of all the colors currently applied, rather than processing individual changes one at a time.